sympy/functions/elementary/piecewise.py

```diff
@@ -1248,4 +1271,205 @@ def piecewise_simplify(expr, **kwargs):
+        d[expr].append(n)
+    # Filter out expressions appearing more than once.
+    candidates = {k: v for (k, v) in d.items() if len(v) >= 2}
+    if candidates:  # Any expression appearing more than once?
+        # Find the two instances with the smallest number of
+        # expressions between them (as this probably will give the smallest
+        # logic expressions).
+        c0 = c1 = minstep = len(args)  # Initialize
+        for k, pos in candidates.items():
+            p = -len(args)  # Store previous index, first time there is no previous index.
+            for v in pos:
+                if v:  # If zero => first item, no previous for sure
+                    newstep = v - p
+                    if minstep > newstep:  # Smaller than earlier best?
+                        minstep = newstep
+                        c0 = p
+                        c1 = v
+                p = v
+        # Compute cummulative condition for bypassing intermediate expressions
+        cumcond = And(*[args[k][1] for k in range(c0 + 1, c1)])
+        lasttrue = (args[-1][1] == S.true)  # Remember if last condition is True
+        # Update condition
+        args[c0] = (args[c0][0],
+                    (args[c0][1] | (~cumcond & args[c1][1])).simplify())
+        del args[c1]
+        # Set final condition to true
+        if lasttrue:
+            args[-1] = (args[-1][0], S.true)
+        # Go again
+        return _piecewise_simplify_identical_expressions(args)
+    return args
+
+def _piecewise_replace_equal_expressions(args):
+    d = defaultdict(list)
+    newargs = []
+    uniqueexpr = []
+    for expr, cond in args:
+        unique = True
+        for uexpr in uniqueexpr:
+            if uexpr.equals(expr):
+                unique = False
+                newargs.append((expr, cond))
+                break
+        if unique:
+            newargs.append((expr, cond))
+    return newargs
+
+
+def _piecewise_simplify_conditions(args, measure, ratio):
+    """
+    Propagate earlier conditions to latter to see if they simplify.
+    """
+    def shorter(newcond, oldcond):
+        return newcond if measure(newcond) < ratio*measure(oldcond) else oldcond
+
+    cummulatedcond = S.false
+    newargs = []
+    prevcond = S.false
+    prevnewcond = S.false
+    for expr, cond in args:
+        newcond = (cond & ~cummulatedcond).simplify()
+        shortcond = shorter(newcond, cond)
+        if shortcond != False and shortcond not in [prevcond, prevnewcond]:
+            newargs.append((expr, shortcond))
+        cummulatedcond = (cummulatedcond | shortcond).simplify()
+        prevcond, prevnewcond = cond, newcond
+    return newargs
+
+
+def piecewise_simplify(expr, **kwargs):
+    """
+    Simplify a Piecewise expression
+    """
+
+    if not isinstance(expr, Piecewise):
+        return expr
+
+    expr = piecewise_simplify_arguments(expr, **kwargs)
+    args = list(expr.args)
+    measure = kwargs['measure']
+    ratio = kwargs['ratio']
```

Comment:
Are these arguments mandatory? This will raise if e.g. ratio isn't provided.

Reply:
The idea is not that this should be an end-user function, but called either from `_eval_simplify` or from `simplify`. Maybe I should rename it to reflect that.